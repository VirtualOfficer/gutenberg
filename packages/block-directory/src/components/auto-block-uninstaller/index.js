/**
 * WordPress dependencies
 */
import { unregisterBlockType } from '@wordpress/blocks';
import { useDispatch, useSelect } from '@wordpress/data';
import { useEffect } from '@wordpress/element';

/**
 * Check if a block list contains a specific block type. Recursively searches
 * through `innerBlocks` if they exist.
 *
 * @param {Object} blockType A block object to search for.
 * @param {Object[]} blocks  The list of blocks to look through.
 *
 * @return {boolean} Whether the blockType is found.
 */
function hasBlockType( blockType, blocks = [] ) {
	if ( ! blocks.length ) {
		return false;
	}
	if ( blocks.some( ( { name } ) => name === blockType.name ) ) {
		return true;
	}
	for ( let i = 0; i < blocks.length; i++ ) {
		if ( hasBlockType( blockType, blocks[ i ].innerBlocks ) ) {
			return true;
		}
	}

	return false;
}

export default function AutoBlockUninstaller() {
	const { uninstallBlockType } = useDispatch( 'core/block-directory' );
	const usedBlocks = useSelect( ( select ) =>
		select( 'core/block-editor' ).getBlocks()
	);

	const shouldRemoveBlockTypes = useSelect( ( select ) => {
		const { isAutosavingPost, isSavingPost } = select( 'core/editor' );
		return isSavingPost() && ! isAutosavingPost();
	} );

	const unusedBlockTypes = useSelect(
		( select ) => {
			const { getInstalledBlockTypes } = select( 'core/block-directory' );
			const installedBlocks = getInstalledBlockTypes();

			const blockList = [];
			installedBlocks.forEach( ( blockType ) => {
				if ( ! hasBlockType( blockType, usedBlocks ) ) {
					blockList.push( blockType );
				}
			} );

			return blockList;
		},
		[ usedBlocks ]
	);

	useEffect( () => {
		if ( shouldRemoveBlockTypes && unusedBlockTypes.length ) {
			unusedBlockTypes.forEach( ( blockType ) => {
				uninstallBlockType( blockType );
				unregisterBlockType( blockType.name );
			} );
		}
	}, [ shouldRemoveBlockTypes ] );

	return null;
}
