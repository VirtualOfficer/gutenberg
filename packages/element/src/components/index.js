export { default as pure } from './pure';
export { default as RawHTML } from './raw-html';
