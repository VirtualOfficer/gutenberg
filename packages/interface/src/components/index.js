export { default as PluginComplementaryArea } from './plugin-complementary-area';
export { default as PluginComplementaryAreaHeader } from './plugin-complementary-area-header';
export { default as PinnedItems } from './pinned-items';
