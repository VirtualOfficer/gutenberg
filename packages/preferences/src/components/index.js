export { default as MoreMenuDropdown } from './more-menu-dropdown';
export { default as MoreMenuFeatureToggle } from './more-menu-feature-toggle';
