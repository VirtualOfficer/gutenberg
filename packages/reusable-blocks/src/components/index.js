export { default as ReusableBlocksMenuItems } from './reusable-blocks-menu-items';
export { default as __experimentalCreatePatternModal } from './create-pattern-modal';
