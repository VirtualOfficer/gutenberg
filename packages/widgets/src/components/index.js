export { default as MoveToWidgetArea } from './move-to-widget-area';
export { default as TemplatePartImportWidgetAreaControl } from './template-part-import-widget-area-control';
