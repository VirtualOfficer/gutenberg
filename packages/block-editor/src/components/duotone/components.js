/**
 * WordPress dependencies
 */
import { SVG } from '@wordpress/components';

/**
 * Internal dependencies
 */
import { __unstableGetValuesFromColors as getValuesFromColors } from './index';


/**
 * Stylesheet for rendering the duotone filter.
 *
 * @param {Object} props          Duotone props.
 * @param {string} props.selector Selector to apply the filter to.
 * @param {string} props.id       Unique id for this duotone filter.
 *
 * @return {WPElement} Duotone element.
 */
export function DuotoneStylesheet( { selector, id } ) {
	const css = `
${ selector } {
	filter: url( #${ id } );
}
`;
	return <style>{ css }</style>;
}

/**
 * Values for the SVG `feComponentTransfer`.
 *
 * @typedef Values {Object}
 * @property {number[]} r Red values.
 * @property {number[]} g Green values.
 * @property {number[]} b Blue values.
 * @property {number[]} a Alpha values.
 */

/**
 * SVG for rendering the duotone filter.
 *
 * @param {Object} props        Duotone props.
 * @param {string} props.id     Unique id for this duotone filter.
 * @param {Values} props.values R, G, B, and A values to filter with.
 *
 * @return {WPElement} Duotone element.
 */
export function DuotoneFilter( { id, values } ) {
	return (
		<SVG
			xmlnsXlink="http://www.w3.org/1999/xlink"
			viewBox="0 0 0 0"
			width="0"
			height="0"
			focusable="false"
			role="none"
			style={ {
				visibility: 'hidden',
				position: 'absolute',
				left: '-9999px',
				overflow: 'hidden',
			} }
		>
			<defs>
				<filter id={ id }>
					<feColorMatrix
						// Use sRGB instead of linearRGB so transparency looks correct.
						colorInterpolationFilters="sRGB"
						type="matrix"
						// Use perceptual brightness to convert to grayscale.
						values="
							.299 .587 .114 0 0
							.299 .587 .114 0 0
							.299 .587 .114 0 0
							.299 .587 .114 0 0
						"
					/>
					<feComponentTransfer
						// Use sRGB instead of linearRGB to be consistent with how CSS gradients work.
						colorInterpolationFilters="sRGB"
					>
						<feFuncR
							type="table"
							tableValues={ values.r.join( ' ' ) }
						/>
						<feFuncG
							type="table"
							tableValues={ values.g.join( ' ' ) }
						/>
						<feFuncB
							type="table"
							tableValues={ values.b.join( ' ' ) }
						/>
						<feFuncA
							type="table"
							tableValues={ values.a.join( ' ' ) }
						/>
					</feComponentTransfer>
					<feComposite
						// Re-mask the image with the original transparency since the feColorMatrix above loses that information.
						in2="SourceGraphic"
						operator="in"
					/>
				</filter>
			</defs>
		</SVG>
	);
}

/**
 * SVG filter from on a duotone preset.
 *
 * @param {Object} props          Preset props.
 * @param {Object} props.preset   Duotone preset.
 *
 * @return {WPElement} Duotone element.
 */
export function PresetDuotoneFilter( { preset } ) {
	return (
		<DuotoneFilter
			id={ `wp-duotone-${ preset.slug }` }
			values={ getValuesFromColors( preset.colors ) }
		/>
	);
}
