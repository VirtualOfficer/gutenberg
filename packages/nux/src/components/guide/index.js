/**
 * External dependencies
 */
import classnames from 'classnames';

/**
 * WordPress dependencies
 */
import { useState, Children } from '@wordpress/element';
import { Modal, KeyboardShortcuts, IconButton } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { withSelect } from '@wordpress/data';

/**
 * Internal dependencies
 */
import PageControl from './page-control';
import { BackButtonIcon, ForwardButtonIcon } from './icons';
import FinishButton from './finish-button';

export function Guide( { children, className, finishButtonText, isMobile, onFinish } ) {
	const [ currentPage, setCurrentPage ] = useState( 0 );

	const numberOfPages = Children.count( children );
	const canGoBack = currentPage > 0;
	const canGoForward = currentPage < numberOfPages - 1;

	const goBack = () => {
		if ( canGoBack ) {
			setCurrentPage( currentPage - 1 );
		}
	};

	const goForward = () => {
		if ( canGoForward ) {
			setCurrentPage( currentPage + 1 );
		}
	};

	if ( numberOfPages === 0 ) {
		return null;
	}

	return (
		<Modal className={ classnames( 'nux-guide', className ) } onRequestClose={ onFinish }>

			<KeyboardShortcuts key={ currentPage } shortcuts={ {
				left: goBack,
				right: goForward,
			} } />

			<div className="nux-guide__container">

				{ children[ currentPage ] }

				{ isMobile && ! canGoForward && (
					<FinishButton onClick={ onFinish }>
						{ finishButtonText || __( 'Finish' ) }
					</FinishButton>
				) }

				<div className="nux-guide__footer">
					{ canGoBack && (
						<IconButton
							className="nux-guide__back-button"
							icon={ <BackButtonIcon /> }
							onClick={ goBack }
						>
							{ __( 'Previous' ) }
						</IconButton>
					) }
					<PageControl
						currentPage={ currentPage }
						numberOfPages={ numberOfPages }
						setCurrentPage={ setCurrentPage }
					/>
					{ canGoForward && (
						<IconButton
							className="nux-guide__forward-button"
							icon={ <ForwardButtonIcon /> }
							onClick={ goForward }
						>
							{ __( 'Next' ) }
						</IconButton>
					) }
					{ ! isMobile && ! canGoForward && (
						<FinishButton
							className="nux-guide__finish-button"
							onClick={ onFinish }
						>
							{ finishButtonText || __( 'Finish' ) }
						</FinishButton>
					) }
				</div>

			</div>

		</Modal>
	);
}

export function Page( props ) {
	return <div { ...props } />;
}

const EnhancedGuide = withSelect( ( select ) => ( {
	isMobile: select( 'core/viewport' ).isViewportMatch( '< small' ),
} ) )( Guide );

EnhancedGuide.Page = Page;

export default EnhancedGuide;
