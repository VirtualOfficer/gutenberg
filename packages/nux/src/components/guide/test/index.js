/**
 * External dependencies
 */
import { shallow } from 'enzyme';

/**
 * WordPress dependencies
 */
import { Modal } from '@wordpress/components';

/**
 * Internal dependencies
 */
import { Guide, Page } from '../';
import PageControl from '../page-control';

describe( 'Guide', () => {
	it( 'renders nothing when there are no pages', () => {
		const wrapper = shallow( <Guide /> );
		expect( wrapper.isEmptyRender() ).toBe( true );
	} );

	it( 'renders when there are pages', () => {
		const wrapper = shallow(
			<Guide>
				<Page>Page 1</Page>
				<Page>Page 2</Page>
			</Guide>
		);
		expect( wrapper ).toMatchSnapshot();
	} );

	it( 'hides back button and shows forward button on the first page', () => {
		const wrapper = shallow(
			<Guide>
				<Page>Page 1</Page>
				<Page>Page 2</Page>
			</Guide>
		);
		expect( wrapper.find( PageControl ).prop( 'currentPage' ) ).toBe( 0 );
		expect( wrapper.find( '.nux-guide__back-button' ) ).toHaveLength( 0 );
		expect( wrapper.find( '.nux-guide__forward-button' ) ).toHaveLength( 1 );
		expect( wrapper.find( '.nux-guide__finish-button' ) ).toHaveLength( 0 );
	} );

	it( 'shows back button and shows finish button on the last page', () => {
		const wrapper = shallow(
			<Guide>
				<Page>Page 1</Page>
				<Page>Page 2</Page>
			</Guide>
		);
		wrapper.find( '.nux-guide__forward-button' ).simulate( 'click' );
		expect( wrapper.find( PageControl ).prop( 'currentPage' ) ).toBe( 1 );
		expect( wrapper.find( '.nux-guide__back-button' ) ).toHaveLength( 1 );
		expect( wrapper.find( '.nux-guide__forward-button' ) ).toHaveLength( 0 );
		expect( wrapper.find( '.nux-guide__finish-button' ) ).toHaveLength( 1 );
	} );

	it( 'shows the finish button inline with the content on mobile', () => {
		const wrapper = shallow(
			<Guide isMobile>
				<Page>Page 1</Page>
			</Guide>
		);
		expect( wrapper.find( '.nux-guide__finish-button' ) ).toHaveLength( 0 );
		expect( wrapper.findWhere( ( node ) => node.text() === 'Finish' ) ).toHaveLength( 1 );
	} );

	it( 'calls onFinish when the finish button is clicked', () => {
		const onFinish = jest.fn();
		const wrapper = shallow(
			<Guide onFinish={ onFinish }>
				<Page>Page 1</Page>
			</Guide>
		);
		wrapper.find( '.nux-guide__finish-button' ).simulate( 'click' );
		expect( onFinish ).toHaveBeenCalled();
	} );

	it( 'calls onFinish when the modal is closed', () => {
		const onFinish = jest.fn();
		const wrapper = shallow(
			<Guide onFinish={ onFinish }>
				<Page>Page 1</Page>
			</Guide>
		);
		wrapper.find( Modal ).prop( 'onRequestClose' )();
		expect( onFinish ).toHaveBeenCalled();
	} );
} );
