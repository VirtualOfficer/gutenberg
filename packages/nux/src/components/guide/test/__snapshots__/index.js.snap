// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Guide renders when there are pages 1`] = `
<WithInstanceId(Modal)
  className="nux-guide"
>
  <KeyboardShortcuts
    key="0"
    shortcuts={
      Object {
        "left": [Function],
        "right": [Function],
      }
    }
  />
  <div
    className="nux-guide__container"
  >
    <Page>
      Page 1
    </Page>
    <div
      className="nux-guide__footer"
    >
      <PageControl
        currentPage={0}
        numberOfPages={2}
        setCurrentPage={[Function]}
      />
      <ForwardRef(IconButton)
        className="nux-guide__forward-button"
        icon={<ForwardButtonIcon />}
        onClick={[Function]}
      >
        Next
      </ForwardRef(IconButton)>
    </div>
  </div>
</WithInstanceId(Modal)>
`;
